Enhance CaffeReader and Caffe2Reader to support multiple LMDB files

00liujj opened the pull request:
Signed-off-by: Jianjun Liu <00liujj@163.com>

#### Why we need this PR?

#1208 

- Refactoring to enhance LMDBLoader so it can support multiple LMDB files.

#### What happened in this PR?
 - Implements a new LMDBLoader.
 - Changes the type of path option of CaffeReader and Caffe2Reader from DALI_STRING to DALI_STRING_VEC, for python it will compatible with previous API.
 - Adds a test 'test_operator_caffe_reader.py'. 
 - 'make lint' passed.

awolant commented:
!build

dali-automaton commented:
**CI MESSAGE**: [939021]: BUILD STARTED

00liujj commented:
What should I do for the failure [939021]?

JanuszL commented on dali/pipeline/operators/reader/loader/lmdb.h:
```diff
@@ -30,65 +31,147 @@ namespace dali {
-    mdb_env_ = nullptr;
+  }
+
+  void MapIndexToFile(Index index, Index* file_index, Index* local_index) {
```
You can pass the variables by the reference.

JanuszL commented on dali/test/python/test_operator_caffe_reader.py:
```diff
@@ -0,0 +1,201 @@
+                assert_array_equal(label[0], db_stat[dstid][1])
+                count+=1
+
+def test_reader_vs_db_stat():
```
I don't know if checking `caffe_db_folder` against the ground truth is not too fragile. Maybe it is enough to online generate the ground truth for `caffe_db_folder` and test it against `[caffe_db_folder], [caffe_db_folder, caffe_db_folder]` later.

00liujj replied:
Thank you for your review. 
Do you mean creating an instance of `CaffeReader(path = caffe_db_folder)` and then testing it against the instance of `CaffeReader(path = [caffe_db_folder])` and `CaffeReader(path = [caffe_db_folder, caffe_db_folder])`?

JanuszL replied:
Yes, that is the idea. But it is not a must, rather an idea.

00liujj replied:
Thank you for your reply. I removed the fragile tests that depend on a specific LMDB file, and added the test as you said.

JanuszL commented on dali/pipeline/operators/reader/loader/lmdb.h:
```diff
@@ -30,65 +31,147 @@ namespace dali {
-    mdb_env_ = nullptr;
+  }
+
+  void MapIndexToFile(Index index, Index* file_index, Index* local_index) {
```
Although it is not used here it is possible to use this code with any form of random access to lmdb. The question is if it is needed?
Maybe we can assume that we just read sequentially entries from the first path, db run out of samples we move to the second path, and at the end, we wrap to the first path back.

00liujj replied:
Thank you for your review. I have implemented a RandomLMDBLoader as a DALI plugin and used it for my caffe detection project. 
It reads from a random dbpath at first, when run out of samples it will randomly chooses the next dbpath to read. 
If it is needed I could add this to the PR.

JanuszL replied:
Random selection of path to reading next won't have any perf implication, I only worry about reading any sample randomly from any db.
If you want to add such a feature you need to make sure that shards doesn't overlap and are consistent between DALI instances running on the different GPUs.

00liujj replied:
It it hard to conform to the shard mechanism for RandomLMDBLoader without any perf implication, it is use for single GPU only and is not suitable for this PR.

JanuszL commented:
!build

JanuszL commented:
LGTM, let us see if CI passes (we have our internal problems with it now as well, so don't worry if it fails for the first time).

dali-automaton commented:
**CI MESSAGE**: [939542]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [939542]: BUILD FAILED

00liujj commented:
@JanuszL Could you tell me whether the failure [939542] is related to this PR? If that, Would you send me the failure log so I can improve the PR.

JanuszL commented:
Let me run it again.

dali-automaton commented:
**CI MESSAGE**: [945319]: BUILD STARTED

JanuszL commented:
I'm afraid that tests are failing because:
```
======================================================================
FAIL: test_operator_caffe_reader.test_reader_vs_simple('/opt/dali_extra/db/lmdb', 44, 1, 1, [array([[[211, 217, 217],
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/opt/dali/dali/test/python/test_operator_caffe_reader.py", line 95, in check_reader_vs_simple
    assert_array_equal(image, images[dstid])
  File "/usr/local/lib/python3.6/dist-packages/numpy/testing/_private/utils.py", line 918, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/usr/local/lib/python3.6/dist-packages/numpy/testing/_private/utils.py", line 841, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

Mismatch: 99.8%
Max absolute difference: 255
Max relative difference: 31.25
 x: array([[[ 58,  27,  24],
        [ 56,  25,  22],
        [ 57,  24,  19],...
 y: array([[[211, 217, 217],
        [213, 219, 219],
        [213, 219, 219],...
```
Could you run this test locally and check the results?

00liujj commented:
@JanuszL Thank you very much for giving me the feedback. I am going to reproduce it.

00liujj commented:
@JanuszL I checked the results and found the problem. I had one more test related to the specific LMDB and my local LMDB is old. I removed it and merged the test logic to one. 

Another problem is my branch has conflicts with master, what should I do?

jantonguirao commented:
@00liujj Those files in the list of conflicts were moved from `dali/pipeline/operators` to `dali/operators` recently so you need to rebase your PR with upstream master and resolve your conflicts. Sorry for the inconvenience

00liujj commented:
@JanuszL Could you please test the PR? I hope it will pass.

JanuszL commented:
!build

JanuszL commented:
@00liujj - I have rerun it. Let us see what the results is now.

dali-automaton commented:
**CI MESSAGE**: [948274]: BUILD STARTED

00liujj commented:
@JanuszL Thank you very much.

dali-automaton commented:
**CI MESSAGE**: [948274]: BUILD FAILED

jantonguirao commented:
@00liujj It is failing on the test `test_pipeline.test_lazy_init_empty_data_path` from `dali/test/python/test_pipeline.py`

00liujj commented:
@jantonguirao Thank you for your help. I have reproduced the segfault and will fix it soon.

JanuszL commented:
> @jantonguirao Thank you for your help. I have reproduced the segfault and will fix it soon.

@00liujj - we really appreciate your effort

00liujj commented:
@jantonguirao @JanuszL I fixed the problem, it's really hard to find that without your experience. I found that the gtest fails and fixed it at the same time.

I finished the following tests:
1. `./dali/python/nvidia/dali/test/dali_test.bin --gtest_filter='Data*'`
2. `nosetests ../dali/test/python/test_pipeline.py -m 'lazy' --verbose`
3. `nosetests ../dali/test/python/test_operator_reshape.py`

Please take a look, thank you!

JanuszL commented:
!build

dali-automaton commented:
**CI MESSAGE**: [949293]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [949293]: BUILD FAILED

00liujj commented:
!build

JanuszL commented:
!build

JanuszL commented:
@00liujj - thanks. Will test it again.

dali-automaton commented:
**CI MESSAGE**: [950441]: BUILD STARTED

00liujj commented on dali/test/python/test_operator_caffe_reader.py:
```diff
@@ -0,0 +1,85 @@
+# with different batch_size and num_threads
+def check_reader_path_vs_paths(paths, batch_size1, batch_size2, num_threads1, num_threads2):
+ 
+    pipe1 = CaffeReaderPipeline(caffe_db_folder, batch_size1, num_threads1)
```
Thank you for your review. I got your mean. The difference is that `compare_pipelines` compares two batches from two pipes, while `test_reader_path_vs_paths` compares two samples from two pipes. `Seq(pipe)` will generate the sample sequence of `pipe`.
